waiter/src/waiter/service_description.clj

```diff
@@ -118,24 +120,36 @@
 ; keys allowed in the token data
 (def ^:const token-data-keys (set/union service-description-keys token-metadata-keys))
 
-(defn transform-allowed-params
-  "Converts allowed-params in the service-description to a set."
+(defn transform-allowed-params-header
+  "Converts allowed-params comma-separated string in the service-description to a set."
   [service-description]
   (cond-> service-description
           (contains? service-description "allowed-params")
           (update "allowed-params"
                   (fn [allowed-params]
                     (cond
-                      (coll? allowed-params)
-                      (set allowed-params)
-
                       (string? allowed-params)
                       (if-not (str/blank? allowed-params)
                         (set (str/split allowed-params #","))
                         #{})
 
                       :else
-                      (throw (ex-info "Provided allowed-params is not a vector or a string"
+                      (throw (ex-info "Provided allowed-params is not a string"
+                                      {:allowed-params allowed-params :status 400})))))))
+
+(defn transform-allowed-params-token-entry
+  "Converts allowed-params vector in the service-description to a set."
+  [service-description]
+  (cond-> service-description
+          (contains? service-description "allowed-params")
+          (update "allowed-params"
+                  (fn [allowed-params]
+                    (cond
```

Comment:
Nitpick: Use `when-not ... throw` pattern.

Reply:
Done.